src/webapi-key-manager.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH WEBAPI-KEY-MANAGER "1" "January 2025" "givemail 1.31" "User Commands"
.SH NAME
webapi-key-manager \- Manage WebAPI keys
.SH SYNOPSIS
.B webapi-key-manager
[\fI\,OPTIONS\/\fR] [\fI\,ID\/\fR]
.SH DESCRIPTION
webapi\-key\-manager \- Manage WebAPI keys
.SH OPTIONS
.HP
\fB\-c\fR, \fB\-\-configuration\-file\fR FILENAME load configuration from the given file
.TP
\fB\-d\fR, \fB\-\-delete\fR
delete the specified key ID
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-n\fR, \fB\-\-new\fR DAYS
generate a new key valid for the given number of days
.TP
\fB\-s\fR, \fB\-\-sign\fR FILENAME
sign the request XML found in the given file with the specified key ID
.TP
\fB\-v\fR, \fB\-\-version\fR
output version information and exit
.PP
This is free software.  You may redistribute copies of it under the terms of
the GNU Lesser General Public License <http://www.gnu.org/licenses/old\-licenses/lgpl\-2.1.html>.
There is NO WARRANTY, to the extent permitted by law.


NEWS
2025/01/27 1.31
* General
  Caught up with changes to MySQL 8.0.

2020/06/20 1.30
* General
  Obsoleted DomainKeys in favour of DKIM with opendkim.
  The GIVEMAIL_VERIFY_SIGNATURES environment variable forces verifications.
  A public key may be provided via the configuration file.
  Miscellaneous fixes.

2015/08/02 1.20
* General
  XML input files may specify useragent/xmailer. The XML document's encoding
  is used as the charset of all message parts.

* libetpan backend
  Initial support for generation of read receipts and delivery notifications.
  Fixed build against libetpan v1.6 and v1.0.
  Fixed some memory leaks.

2014/05/19 1.11
* General
  No suffix is appended to message IDs passed with -m.
  Content and attachments' default encoding may be overriden in XML input.
  Copes with older libetpan releases that don't support NTLM and GSSAPI.
  Fixed file sizes and offsets to use off_t.

2012/11/29 1.10
* General
  Require ctemplate.
  XML input files loaded with givemail -x that advertise version 2 may use ctemplate
  markers and sections.

* libetpan backend
  Enable libetpan mail stream debugging if GIVEMAIL_DEBUG=Y.
  On authentication errors, set recipients status to -2.
  If an error occurs, get the SASL error message.
  Fixed embedding of text and HTML parts with file attachments.
  Prefer LOGIN and PLAIN to MD5 auth methods.
  Always set a Date header.

2011/12/07 1.06
* General
  If master/msgidsuffix isn't in the configuration file, it defaults to "@hostname".

* libetpan backend
  SMTP authentication with NTLM.
  Fixed QP-encoding.

* givemail
  Option -d/--dump-file instructs givemail to write to disk the plain text and
  HTML parts of messages loaded with -x/--xml-file, as sent to recipients. Files
  are keyed by message ID, with extension txt or html.
  Message IDs are saved to the status file.

2011/07/03 1.05
* General
  ltmain.sh supports --as-needed.
  In the configuration file, master/sendersuffix and slave/smtpprovider were obsoleted.
  Refactored code to bundle the mail sending core in a dynamic library.

* libetpan backend
  QP-encoding of non-ASCII characters applies to separate words, not whole strings.
  Catch SASL authentication errors, prefer MD5 methods to PLAIN, PLAIN to LOGIN.

* SMTP
  Allow setting the mail relay as a hostname or an IP address.
  Inline attachments may be skipped if not referenced within the HTML content.
  
* csv2givemail
  Don't complain if there is no file name pattern to match.

* givemail
  The file pointed to by -x may specify a To, regardless of the recipients list.
  Custom fields may have any name.
  The -f option was fixed.

* givemaild
  Dropped the unloved legacy DB schema (new schema not stabilized).

* WebAPI
  When importing a CSV, some fields may span several columns.
  Trim spaces around email addresses.
  Allow more recipient properties to be imported.

2010/05/11 1.04
* SMTP
  In the configuration file, relay/server was dropped and replaced with relay/address and relay/port.
  The environment variable GIVEMAIL_RELAY_SERVER is obsolete. Use GIVEMAIL_RELAY_ADDRESS and GIVEMAIL_RELAY_PORT.

2010/03/10 1.03
* SMTP
  DSN can be activated by setting slave/dsnnotify in the configuration file.

* libetpan backend
  Names with non-ASCII characters are QP-encoded, assumed to be UTF-8.
  Fixed handling of email address lists.
  Fixed session shutdown.

* givemail
  Recipients' SMTP statuses can be saved to the file specified with -t/--status-file.
  In -x/--xml-file mode, a CC header may be specified.

2009/12/13 1.02
* SMTP
  All code that depends on libesmtp was moved to a backend.
  A new backend based on libetpan is available.
  It's possible to have attachments inline within the HTML content.
  SMTP relay information can be provided through environment variables (GIVEMAIL_RELAY_*).
  Subject, From, Reply-To, Sender and Resent-From can have substitution fields.
  The number of substitution fields isn't artificially limited to 6 any longer.

* libetpan backend
  STARTTLS can be enabled when sending through the relay.

* givemail
  When run with the -x parameter, the message ID can be set with -m. Replies can be generated with -e.

2009/09/08 1.01
* General
  The database schema changed. The recipients table has a timesent and a numattempts column.
  All items in the configuration file are documented.
  Fixed issues reported by gcc 4.4.

* SMTP
  Authentication against the SMTP relay is supported. Username and password can be set in givemail.conf.
  DSN was activated.
  Substitution fields in the display name part of From are supported.

* csv2givemail
  It looks for CCs in the column immediately behind the To recipients' column, and for BCCs in the column that follows.
  If -b/--base-directory isn't set, it defaults to the current working directory.

* givemail
  The SMTP relay configured in givemail.conf is taken in consideration when run in slave mode.
  When run with the -x parameter, values of substitution fields can be set on the command-line, or via another file passed with the -f parameter.

* givemaild
  The daemon tries to keep to a minimum the number of recipients that failed with the libesmtp error "Invalid API function argument". The root cause of this error has not been identified yet.
  Every hour or so, it looks for campaigns with temporarily failed recipients, between one and three days old, and reactivates those campaigns.

* WebAPI
  It returns and can set most properties on campaigns and recipients.
  Recipients can be imported from a CSV file (name, address, status), or from another campaign.
  Campaigns lists of recipients can be exported to CSV.
  Recipients' status can be updated en masse.
  Lists of campaigns and recipients can be obtained with simple GET requests.
  It can return lists of objects changed since a given time.
  Duplicate recipients cannot be added.
  The authentication scheme relies on cookies.
  Hosts for which authentication can be skipped are defined in the configuration file.

src/webapi-key-manager.cc
/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 4; tab-width: 4 -*- */
/*
 *  Copyright 2008 Global Sign In
 *  Copyright 2009-2021 Fabrice Colin
 * 
 *  This code is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <getopt.h>
#include <signal.h>
#include <time.h>
#include <libintl.h>
#include <iostream>
#include <fstream>
#include <algorithm>

#include "config.h"
#include "AuthSQL.h"
#include "Base64.h"
#include "ConfigurationFile.h"
#include "HMAC.h"
#include "MessageDetails.h"
#include "MySQLBase.h"
#include "TimeConverter.h"

using namespace std;

static bool g_mustQuit = false;

static struct option g_longOptions[] = {
	{"configuration-file", required_argument, NULL, 'c'},
	{"delete", no_argument, NULL, 'd'},
	{"help", no_argument, NULL, 'h'},
	{"new", required_argument, NULL, 'n'},
	{"sign", required_argument, NULL, 's'},
	{"version", no_argument, NULL, 'v'},
	{0, 0, 0, 0}
};

/// Prints an help message.
static void printHelp(void)
{
	// Help
	cout << "webapi-key-manager - Manage WebAPI keys\n\n"
		<< "Usage: webapi-key-manager [OPTIONS] [ID]\n\n"
		<< "Options:\n"
		<< "  -c, --configuration-file FILENAME load configuration from the given file\n"
		<< "  -d, --delete                      delete the specified key ID\n"
		<< "  -h, --help                        display this help and exit\n"
		<< "  -n, --new DAYS                    generate a new key valid for the given number of days\n"
		<< "  -s, --sign FILENAME               sign the request XML found in the given file with the specified key ID\n"
		<< "  -v, --version                     output version information and exit" << endl;
}

/// Catch signals and take the appropriate action.
static void catchSignals(int sigNum)
{
	if (sigNum != SIGCHLD)
	{
		g_mustQuit = true;
	}
}

/// Get some random bytes.
static char *getRandomBytes(off_t bytesCount)
{
	char *pBuffer = NULL;
	off_t bytesRead = 0;

	if (bytesCount == 0)
	{
		return NULL;
	}

	ifstream randomFile("/dev/random");

	if (randomFile.is_open() == true)
	{
		pBuffer = new char[bytesCount + 1];

		while (bytesRead < bytesCount)
		{
			randomFile.read(pBuffer, bytesCount - bytesRead);
			bytesRead += randomFile.gcount();
#ifdef DEBUG
			cout << "Read " << bytesRead << " bytes" << endl;
#endif
		}
		pBuffer[bytesCount] = '\0';

		randomFile.close();
	}

	return pBuffer;
}

int main(int argc, char **argv)
{
	struct sigaction quitAction, ignAction;
	string configFileName("/etc/givemail/conf.d/givemail.conf"), requestFileName, keyId;
	int longOptionIndex = 0, minimumArgsCount = 0;
	int returnCode = EXIT_FAILURE;
	unsigned int daysCount = 30;
	bool generateKey = false, deleteKey = false, signRequest = false;

	// Look at the options
	int optionChar = getopt_long(argc, argv, "c:dhn:s:v", g_longOptions, &longOptionIndex);
	while (optionChar != -1)
	{
		switch (optionChar)
		{
			case 'c':
				if (optarg != NULL)
				{
					configFileName = optarg;
				}
				break;
			case 'd':
				minimumArgsCount = 1;
				deleteKey = true;
				break;
			case 'h':
				printHelp();
				return EXIT_SUCCESS;
			case 'n':
				if (optarg != NULL)
				{
					daysCount = (unsigned int)atoi(optarg);
					generateKey = true;
				}
				break;
			case 's':
				if (optarg != NULL)
				{
					requestFileName = optarg;
					minimumArgsCount = 1;
					signRequest = true;
				}
				break;
			case 'v':
				cout << "webapi-key-manager - " << PACKAGE_STRING << "\n\n"
					<< "This is free software.  You may redistribute copies of it under the terms of\n"
					<< "the GNU Lesser General Public License <http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html>.\n"
					<< "There is NO WARRANTY, to the extent permitted by law." << endl;
				return EXIT_SUCCESS;
			default:
				return EXIT_FAILURE;
		}

		// Next option
		optionChar = getopt_long(argc, argv, "c:dhn:s:v", g_longOptions, &longOptionIndex);
	}

	if ((argc - optind < minimumArgsCount) ||
		((generateKey == false) && (deleteKey == false) && (signRequest == false)))
	{
		printHelp();
		return EXIT_FAILURE;
	}

	if (minimumArgsCount >= 1)
	{
		keyId = argv[optind];
	}

#if defined(ENABLE_NLS)
	bindtextdomain(GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
	bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
	textdomain(GETTEXT_PACKAGE);
#endif

	// Catch SIGINT, QUIT and CHLD
	sigemptyset(&quitAction.sa_mask);
	quitAction.sa_flags = 0;
	quitAction.sa_handler = catchSignals;
	sigaction(SIGINT, &quitAction, NULL);
	sigaction(SIGQUIT, &quitAction, NULL);
	sigaction(SIGCHLD, &quitAction, NULL);
	sigemptyset(&ignAction.sa_mask);
	// Ignore SIGPIPE
	ignAction.sa_flags = 0;
	ignAction.sa_handler = SIG_IGN;
	sigaction(SIGPIPE, &ignAction, NULL);

	// Load configuration
	ConfigurationFile *pConfig = ConfigurationFile::getInstance(configFileName);
	if ((pConfig == NULL) ||
		(pConfig->parse() == false))
	{
		cerr << "Couldn't open configuration file " << configFileName << endl;
		return EXIT_FAILURE;
	}

	// Open the database
	MySQLBase db(pConfig->m_hostName, pConfig->m_databaseName,
		pConfig->m_userName, pConfig->m_password);
	if (db.isOpen() == false)
	{
		return EXIT_FAILURE;
	}

	AuthSQL authData(&db);

	if (generateKey == true)
	{
		Key key;
		unsigned long keyLength = 64;

		key.m_creationTime = time(NULL);
		key.m_expiryTime = key.m_creationTime + (daysCount * 24 * 3600);

		char *pRandomKey = getRandomBytes((off_t)keyLength);
		if (pRandomKey != NULL)
		{
			char *pEncodedKey = Base64::encode(pRandomKey, keyLength);
			if (pEncodedKey != NULL)
			{
				key.m_appId = db.getUniversalUniqueId();
				key.m_value.append(pEncodedKey, keyLength);

				if (authData.createNewKey(key) == true)
				{
					cout << "Generated a new HMAC key" << endl;
					cout << "Key ID:\t\t" << key.m_appId << endl;
					cout << "Key value:\t" << pEncodedKey << endl;
					cout << "Creation time:\t" << key.m_creationTime << endl;
					cout << "Expiry time:\t" << key.m_expiryTime << endl;
					cout << "Forward this information to customer through an appropriately secure channel" << endl;

					returnCode = EXIT_SUCCESS;
				}
				else
				{
					cerr << "Couldn't generate a new HMAC key" << endl;
				}
				delete[] pEncodedKey;
			}

			delete[] pRandomKey;
		}
	}
	else if (deleteKey == true)
	{
		Key *pKey = authData.getKey(keyId);
		if (pKey == NULL)
		{
			cout << "No such HMAC key ID" << endl;
		}
		else
		{
			if (authData.deleteKey(pKey, true) == true)
			{
				cout << "Deleted HMAC key ID " << keyId << endl;

				returnCode = EXIT_SUCCESS;
			}
			else
			{
				cerr << "Couldn't delete HMAC key" << endl;
			}

			delete pKey;
		}
	}
	else if (signRequest == true)
	{
		time_t timeNow = time(NULL);

		Key *pKey = authData.getKey(keyId);
		if (pKey == NULL)
		{
			cout << "No such HMAC key ID" << endl;
		}
		else
		{
			off_t fileSize = 0;

			double keyDiff = pKey->getTimeDiff(timeNow);
			if (keyDiff < 0)
			{
				// Warn and proceed
				cout << "Warning: key has expired" << endl;
			}

			char *pRequest = MessageDetails::loadRaw(requestFileName, fileSize);
			if (pRequest == NULL)
			{
				cout << "Couldn't open file " << requestFileName << endl;
			}
			else
			{
				string dateHeaderValue(TimeConverter::toTimestamp(timeNow, false));
				string payload(dateHeaderValue);

				payload += "\n";
				payload += pRequest;

				off_t dataLength = (off_t)payload.length();

				cout << "Signing " << dataLength << " bytes" << endl;

				char *pEncodedSignature = HMAC::sign(pKey, payload.c_str(), dataLength);
				if (pEncodedSignature != NULL)
				{
					cout << "Prepend the following headers:" << endl;
					cout << "Date: " << dateHeaderValue << endl;
					cout << "Set-Cookie: GMAuth=" << keyId << ":" << pEncodedSignature
						<< "; expires=" << TimeConverter::toTimestamp(pKey->m_expiryTime, true)
						<< "; path=/" << endl << endl;
					cout << "Post to the WebAPI with:" << endl;
					cout << "curl --data-binary @" << requestFileName
						<< " http://localhost/cgi-bin/webapi.fcgi --cookie \"GMAuth="
						<< keyId << ":" << pEncodedSignature << "\"" << endl;

					delete[] pEncodedSignature;
				}

				delete[] pRequest;
			}

			delete pKey;
		}
	}

	return returnCode;
}

